File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.py

from typing import Generator
from collections import deque
import sys
input = sys.stdin.readline


def solve(graph:dict, N, K, X)->Generator[int]:
    '''
    output:
        generates answers as individual yield.
    '''
    que = deque()
    dist = [-1] * (N+1)

    que.append(X)
    dist[X] = 0

    while que:
        now = que.popleft()
        if dist[now] == K:
            yield now
            continue
        if dist[now] > K:
            return
        if now not in graph: # no next node
            continue
        for nxt in graph[now]:
            if dist[nxt] >= 0: # already visited
                continue
            que.append(nxt)
            dist[nxt] = dist[now]+1
    # note: yield nothing if no appropriate answer exist.
    return


N,M,K,X = map(int, input().split())

graph = {}
for _ in range(M):
    a,b = map(int, input().split())
    if a not in graph:
        graph[a] = [b]
    else:
        graph[a].append(b)

ans = [ k for k in solve(graph, N, K, X) ]
if not ans: ans = [-1]
print('\n'.join( map(str, sorted(ans)) ))

File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.cc

#include <iostream>
#include <algorithm>
#include <map>
#include <deque>
#include <sstream>


using namespace std;

/*
    custom graph type.
    key: start node
    value: end node list as vector<int>
*/
using Map = map<int,vector<int>>;

vector<int> solve(Map& map, int N, int K, int X)
{
    auto que = deque<int>();
    auto mindist = vector<int>(N+1, -1); // [0] is not used.
    vector<int> answer;

    // now, start from node X.
    que.push_back(X);
    mindist[X] = 0;

    int cur; // visiting node
    int dist; // distance up to here

    while (!que.empty()) {
        cur = que.at(0);
        dist = mindist[cur];
        que.pop_front();

        if (dist == K) {
            answer.push_back(cur);
            continue; // no need to consider next node.
        }
        if (dist > K) // no more need to continue.
            break;
        if (!map.contains(cur)) // map does not have node-cur.
            continue;
        for (auto nxt: map[cur]) {
            if (mindist[nxt] >= 0)
                continue;
            que.push_back(nxt);
            mindist[nxt] = dist + 1;
        }
    }
    if (answer.empty())
        answer.push_back(-1);

    sort(answer.begin(), answer.end());
    return answer;
}


int main()
{
    int N, M, K, X, A, B;
    scanf("%d%d%d%d", &N, &M, &K, &X);

    auto graph = Map();
    for (int i=0; i<M; i++) {
        scanf("%d%d", &A, &B);
        graph[A].push_back(B);
    }
    auto ans = solve(graph, N, K, X);
    for (auto val: ans)
        printf("%d\n", val);
    return 0;
}
